Add Grafana dashboard and installation steps (#1620)
nginxinc/nginx-gateway-fabric: site/content/how-to/monitoring/prometheus.md

{{<call-out "important" "Security note for metrics">}}
Metrics are served over HTTP by default. Enabling HTTPS will secure the metrics endpoint with a self-signed certificate. When using HTTPS, adjust the Prometheus Pod scrape settings by adding the `insecure_skip_verify` flag to handle the self-signed certificate. For further details, refer to the [Prometheus documentation](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tls_config).
{{</call-out>}}

## Installing Prometheus and Grafana

### Prometheus

```shell
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm install prometheus prometheus-community/prometheus -n monitoring --create-namespace --set server.global.scrape_interval=15s
```

Once running, you can access the Prometheus dashboard by using port-forwarding:

```shell

> kubectl -n monitoring port-forward svc/prometheus-server 9090:80

Reviewer: Can we background this so we don't have to switch shell?
```suggestion
kubectl -n monitoring port-forward svc/prometheus-server 9090:80 &
```

Author: Not sure why, but on my machine it's still running in the foreground...

```

Visit [http://127.0.0.1:9090](http://127.0.0.1:9090) to view the dashboard.

### Grafana

```shell
helm repo add grafana https://grafana.github.io/helm-charts
helm repo update
helm install grafana grafana/grafana -n monitoring --create-namespace
```

Once running, you can access the Grafana dashboard by using port-forwarding:

```shell
kubectl -n monitoring port-forward svc/grafana 3000:80
```

> Visit [http://127.0.0.1:3000](http://127.0.0.1:3000) to view the Grafana UI.

Reviewer: Can we add a note about the username and password here for login? 
Username: admin
Password: `kubectl get secret --namespace monitoring grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo`

#### Configuring Grafana

In the Grafana UI menu, go to `Connections` then `Data sources`. Add your Prometheus service (`http://prometheus-server.monitoring.svc`) as a data source.

Download the following sample dashboard and Import as a new Dashboard in the Grafana UI.

> {{< fa "download" >}} {{< link "static/grafana-dashboard.json" "`ngf-grafana-dashboard.json`" >}}

Author: This isn't quite working. Not sure what the exact syntax/location should be.

Reviewer 2: Removing "static" from the path should do it. The shortcode is using the `relurl`, which is the path relative to the base path in the Hugo config. Hugo doesn't include the static directory in the path when it builds the site, so the url for the file in `static/grafana-dashboard.json` would be `<baseurl>/grafana-dashboard.json`.

Author: Thanks Jodie. That opened the json file properly in another tab, but I expected it to download automatically to my local machine.

Reviewer 2: I need to make an update to the shortcode to tell it to download the file. This code won't have to change to support that behavior, so this is ok to commit as-is. I will make the change and give you a test theme version of the theme to try out.

Reviewer 2: Actually, on second thought, I need to add a new shortcode for this. Changing how the link shortcode works to default to downloading could break stuff elsewhere.

Author: It use to behave by downloading. I know we have other docs that work like this, though I'm sure they are out of date with the latest stuff.

Reviewer 2: @sjberman I created a new alpha release in the nginx-hugo-theme repo that you can use for testing:  https://github.com/nginxinc/nginx-hugo-theme/releases/tag/v0.41.0-alpha-1 

I updated your feature branch settings in netlify to use the alpha version already; if you want to use it locally, you'll need to run `hugo mod get https://github.com/nginxinc/nginx-hugo-theme v0.41.0-alpha-1` from the `site` directory.

If this works out ok, we'll merge it in and roll out a theme update for it. 

The syntax to use for the new shortcode is:

```
{{< download "static/grafana-dashboard.json" "grafana-dashboard.json" >}}
```

(I included the download icon in the shortcode, so you don't need to use the `fa` shortcode with this one.

Author: @jputrino Had some issues updating hugo locally, but I pushed the change. In the Netlify preview, the font is a bit small for the link, and it just opens a blank New Tab when clicking.

Reviewer 2: @sjberman What browser are you using? DId you check to see if the file also downloaded? 
It worked fine for me in Safari, but I haven't tested other browsers yet.

Author: Firefox, wasn't downloaded.

Author: Apparently doesn't work on Chrome either.

Reviewer 2: Thanks for that! We'll have to take a look and see what's up there. I don't want this to be a blocker in the meantime, so I'd say you can either stick with the `file` shortcode, provide a link directly to the file in Github, or create a Gist from it and embed that in the doc. Hugo has built-in support for gists: https://gohugo.io/content-management/shortcodes/#gist

Author: What does the `file` shortcode do?

## Overview

> NGINX Gateway Fabric metrics are displayed in [Prometheus](https://prometheus.io/) format, simplifying monitoring. These metrics are served via a metrics server orchestrated by the controller-runtime package on HTTP port `9113`. When installed, Prometheus automatically scrapes this port and collects metrics. [Grafana](https://grafana.com/) can be used for rich visualization of these metrics.

Reviewer 3: ```suggestion
NGINX Gateway Fabric metrics are displayed in [Prometheus](https://prometheus.io/) format. These metrics are served through a metrics server orchestrated by the controller-runtime package on HTTP port `9113`. When installed, Prometheus automatically scrapes this port and collects metrics. [Grafana](https://grafana.com/) can be used for rich visualization of these metrics.
```

"simplifying" requires qualification: Prometheus and Grafana are conventional choices due to reliability and ubiquity, which do not need to be justified, so there's no need to say anything at all.

We're trying to avoid Latinisms, though it's challenging to identify all of them.

{{<call-out "important" "Security note for metrics">}}
Metrics are served over HTTP by default. Enabling HTTPS will secure the metrics endpoint with a self-signed certificate. When using HTTPS, adjust the Prometheus Pod scrape settings by adding the `insecure_skip_verify` flag to handle the self-signed certificate. For further details, refer to the [Prometheus documentation](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tls_config).
{{</call-out>}}

> ## Installing Prometheus and Grafana

Reviewer 4: Would it make sense to add a disclaimer here that installation instructions for Prometheus and Helm are for demonstration purposes only or something like that?  So that we don't document any production ready deployment of those things.

```shell
kubectl port-forward -n monitoring svc/prometheus-server 9090:80 &
```

Visit [http://127.0.0.1:9090](http://127.0.0.1:9090) to view the dashboard.

### Grafana

```shell
helm repo add grafana https://grafana.github.io/helm-charts
helm repo update
helm install grafana grafana/grafana -n monitoring --create-namespace
```

Once running, you can access the Grafana dashboard by using port-forwarding in the background:

```shell
kubectl port-forward -n monitoring svc/grafana 3000:80 &
```

Visit [http://127.0.0.1:3000](http://127.0.0.1:3000) to view the Grafana UI.

The username for login is `admin`. The password can be acquired by running:

```shell
kubectl get secret -n monitoring grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo
```

#### Configuring Grafana

In the Grafana UI menu, go to `Connections` then `Data sources`. Add your Prometheus service (`http://prometheus-server.monitoring.svc`) as a data source.

Download the following sample dashboard and Import as a new Dashboard in the Grafana UI.

{{< fa "download" >}} {{< link "grafana-dashboard.json" "`ngf-grafana-dashboard.json`" >}}

## Available metrics in NGINX Gateway Fabric

> NGINX Gateway Fabric provides a variety of metrics to assist in monitoring and analyzing performance. These metrics are categorized as follows:

Reviewer 3: ```suggestion
NGINX Gateway Fabric provides a variety of metrics for monitoring and analyzing performance. These metrics are categorized as follows:
```

### NGINX/NGINX Plus metrics

> NGINX metrics, essential for monitoring specific NGINX operations, include details like the total number of accepted client connections. For a complete list of available NGINX/NGINX Plus metrics, refer to the [NGINX Prometheus Exporter developer docs](https://github.com/nginxinc/nginx-prometheus-exporter#exported-metrics).

Reviewer 3: ```suggestion
NGINX metrics cover specific NGINX operations such as the total number of accepted client connections. For a complete list of available NGINX/NGINX Plus metrics, refer to the [NGINX Prometheus Exporter developer docs](https://github.com/nginxinc/nginx-prometheus-exporter#exported-metrics).
```

> These metrics use  the `nginx_gateway_fabric` namespace and include the `class` label, indicating the NGINX Gateway class. For example, `nginx_gateway_fabric_connections_accepted{class="nginx"}`.

Reviewer 3: ```suggestion
These metrics use the `nginx_gateway_fabric` namespace and include the `class` label, indicating the NGINX Gateway class. For example, `nginx_gateway_fabric_connections_accepted{class="nginx"}`.
```

### NGINX Gateway Fabric metrics

> Metrics specific to the NGINX Gateway Fabric include:

Reviewer 3: ```suggestion
Metrics specific to NGINX Gateway Fabric include:
```

Any usage of "the" prepending software names suggests a specific instance, whereas this is describing it in a generic sense.

- `nginx_reloads_total`: Counts successful NGINX reloads.
- `nginx_reload_errors_total`: Counts NGINX reload failures.
- `nginx_stale_config`: Indicates if NGINX Gateway Fabric couldn't update NGINX with the latest configuration, resulting in a stale version.
- `nginx_last_reload_milliseconds`: Time in milliseconds for NGINX reloads.
- `event_batch_processing_milliseconds`: Time in milliseconds to process batches of Kubernetes events.

All these metrics are under the `nginx_gateway_fabric` namespace and include a `class` label set to the Gateway class of NGINX Gateway Fabric. For example, `nginx_gateway_fabric_nginx_reloads_total{class="nginx"}`.

### Controller-runtime metrics

> Provided by the [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) library, these metrics cover a range of aspects:

Reviewer 3: ```suggestion
Provided by the [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) library, these metrics include:
```

```

Once running, you can access the Prometheus dashboard by using port-forwarding in the background:

```shell
kubectl port-forward -n monitoring svc/prometheus-server 9090:80 &
```

Visit [http://127.0.0.1:9090](http://127.0.0.1:9090) to view the dashboard.

### Grafana

```shell
helm repo add grafana https://grafana.github.io/helm-charts
helm repo update
helm install grafana grafana/grafana -n monitoring --create-namespace
```

Once running, you can access the Grafana dashboard by using port-forwarding in the background:

```shell
kubectl port-forward -n monitoring svc/grafana 3000:80 &
```

Visit [http://127.0.0.1:3000](http://127.0.0.1:3000) to view the Grafana UI.

The username for login is `admin`. The password can be acquired by running:

```shell
kubectl get secret -n monitoring grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo
```

#### Configuring Grafana

In the Grafana UI menu, go to `Connections` then `Data sources`. Add your Prometheus service (`http://prometheus-server.monitoring.svc`) as a data source.

Download the following sample dashboard and Import as a new Dashboard in the Grafana UI.

> {{< download "static/grafana-dashboard.json" "ngf-grafana-dashboard.json" >}}

Reviewer 5: The first parameter here should be just `grafana-dashboard.json` without `static/`.
See https://gohugo.io/content-management/static-files/ - Hugo seems to map anything in `/static` to the root of the server.

Reviewer 5: I've added a fix for that on the hugo theme side. I'll update your theme version on the netlify preview and try it.

Reviewer 5: @sjberman https://deploy-preview-1620--nginx-gateway-fabric.netlify.app/nginx-gateway-fabric/how-to/monitoring/prometheus/ is updated now, I think it's looking good now 👍

Author: awesome, thank you! Can this hugo update be merged in and updated in our repo? I don't want to merge this PR and have the docs rendering fail for anyone else because we aren't yet using the latest version.

Reviewer 5: Yes, I'll get the theme change approved, merged, and updated. Will comment here when it's good to merge.

Reviewer 5: The theme has been updated, you should be good to merge.

Author: @nginx-jack Thanks! Do I need to update the hugo theme in this repo as part of this PR?

Reviewer 5: Yeah, it's worth updating `site/go.mod` for local development.
The latest is;
`require github.com/nginxinc/nginx-hugo-theme v0.41.0`

> {{< download "static/grafana-dashboard.json" "ngf-grafana-dashboard.json" >}}

Reviewer 5: ```suggestion
{{< download "grafana-dashboard.json" "ngf-grafana-dashboard.json" >}}
```
